sdk/src/test/java/com/qonversion/android/sdk/internal/user/controller/UserControllerTest.kt

```diff
@@ -0,0 +1,206 @@
+            coVerifyOrder {
+                mockUserCacher.getActual()
+                mockUserService.obtainUserId()
+                mockUserService.getUser(userId)
+                userController.storeUser(mockUser)
+            }
+        }
+
+        @Test
+        fun `get user from cache with error state`() = runTest {
+            // given
+            userController = spyk(userController)
+
+            every {
+                mockUserCacher.getActual()
+            } returns null
+            val exception = QonversionException(ErrorCode.BackendError)
+            coEvery {
+                mockUserService.getUser(userId)
+            } throws exception
+
+            every {
+                mockUserCacher.getActual(CacheState.Error)
+            } returns mockUser
+
+            // when
+            val result = userController.getUser()
+
+            // then
+            assertThat(result).isEqualTo(mockUser)
+            verify(exactly = 0) {
+                mockLogger.info(any())
+                userController.storeUser(any())
+            }
+
+            coVerifyOrder {
+                mockUserCacher.getActual()
+                mockUserService.obtainUserId()
+                mockUserService.getUser(userId)
+                mockLogger.error("Failed to get User from API", exception)
+                mockUserCacher.getActual(CacheState.Error)
+            }
+        }
+
+        @Test
+        fun `user is missing`() = runTest {
+            // given
+            every {
+                mockUserCacher.getActual()
+            } returns null
+
+            coEvery {
+                mockUserService.getUser(userId)
+            } throws QonversionException(ErrorCode.BackendError)
+
+            every {
+                mockUserCacher.getActual(CacheState.Error)
+            } returns null
+
+            // when and then
+            coAssertThatQonversionExceptionThrown(ErrorCode.UserInfoIsMissing) {
+                userController.getUser()
+            }
+
+            verify(exactly = 0) {
+                mockLogger.info(any())
+                userController.storeUser(any())
+            }
+
+            coVerifySequence {
+                mockUserCacher.getActual()
+                mockUserService.obtainUserId()
+                mockUserService.getUser(userId)
+                mockUserCacher.getActual(CacheState.Error)
+            }
+        }
+    }
+
+    @Nested
+    inner class StoreUserTest {
+        @Test
+        fun `store user successful`() {
+            // given
+            val mockUser = mockk<User>()
+            val userSlot = slot<User>()
+            every {
+                mockUserCacher.store(capture(userSlot))
+            } just Runs
+
+            // when
+            userController.storeUser(mockUser)
+
+            // then
+            verify(exactly = 1) { mockUserCacher.store(mockUser) }
+            assertThat(userSlot.captured).isEqualTo(mockUser)
```

Comment:
we can not capture user here as we already check above the call with the expected argument